src/Simplify_Exprs.cpp

```diff
@@ -53,6 +53,36 @@ Expr Simplify::visit(const Broadcast *op, ExprInfo *bounds) {
 
 Expr Simplify::visit(const VectorReduce *op, ExprInfo *bounds) {
     Expr value = mutate(op->value, bounds);
+    if (const Shuffle *shuffle = value.as<Shuffle>()) {
```

Comment:
I find the control flow here quite confusing. Is the idea just that sometimes you can move the vector reduce inside the shuffle, and sometimes that makes the shuffle go away? If so I think this is more general than is_broadcast, and you can do away with is_broadcast:

Given a reduction factor f, I think you can move the reduce inside the shuffle if indices[i] == (indices[i - i%f]/f) * f + (i % f)

E.g. if you reduce by a factor of four, and the shuffle has indices [4 5 6 7 0 1 2 3 12 13 14 15] then you can move the reduce inside the shuffle, and rewrite the shuffle indices to be every fourth one divided by four: [1 0 3]

In the case where the shuffle indices end up being [0 1 2 3 ... lanes - 1], you don't bother adding the shuffle.

This is different to what we discussed, which was that for min/max/and/or total reductions (output lanes = 1) you can dedup and sort the indices of the inner shuffle, and this is a good idea if it results in a nicer indexing pattern (e.g. a slice)

Reply:
Oh I guess the first thing I wrote doesn't handle a total reduction of a shuffle with indices [0 1 0 1 0 1 0 1]. I'm still confused about this though. Can you add a worked example to the comments maybe?

Reply:
I think the general idea of what you described would be to check if each of the indices[i/f...i/f + f - 1] is a permutation of indices that are aligned to a multiple of f. If so, you can move the shuffle outside, downsample the indices by f, and divide them by f.

This is more complicated/general, but might actually be easier to understand than my current less general algorithm. Let me try it and see if I can implement it.

Reply:
Here is an implementation of what you proposed (I think): https://github.com/halide/Halide/compare/simplify-vectorreduce-shuffles2?expand=1

However, this is really a different thing than what this PR implements.

I'm also not sure if we should do this. I'm concerned it will break code that would be faster because the schedule was trying to hit a particular pattern.

I have another different PR that I think does what this PR does better: #5372